nexus/db-queries/src/db/queries/volume.rs

```diff
@@ -0,0 +1,563 @@
+
+        Ok(())
+    }
+}
+
+impl Expression for ConditionallyUpdateVolume {
+    type SqlType = diesel::sql_types::Array<db::model::Volume>;
+}
+
+impl<GroupByClause> ValidGrouping<GroupByClause> for ConditionallyUpdateVolume {
+    type IsAggregate = is_aggregate::Never;
+}
+
+/// Produces a query fragment that will
+///
+/// 1. decrease the number of references for each region snapshot that
+///    a volume references
+/// 2. soft-delete the volume
+/// 3. record the resources to clean up as a serialized CrucibleResources
+///    struct in volume's `resources_to_clean_up` column.
+///
+/// The output should look like:
+///
+/// ```sql
+///  with UPDATED_REGION_SNAPSHOTS_TABLE as (
+///    UPDATE region_snapshot <ConditionallyDecreaseReferences>
+///  ),
+///  REGION_SNAPSHOTS_TO_CLEAN_UP_TABLE as (
+///    <RegionSnapshotsToCleanUp>
+///  ),
+///  UPDATED_VOLUME_TABLE as (
+///    UPDATE volume <ConditionallyUpdateVolume>
+///  )
+///  select case
+///   when volume.resources_to_clean_up is not null then volume.resources_to_clean_up
+///   else (select resources_to_clean_up from UPDATED_VOLUME_TABLE where id = '<volume id>')
+///  end
+///  from volume where id = '<volume id>';
+/// ```
+#[must_use = "Queries must be executed"]
+#[derive(Debug, Clone)]
+pub struct DecreaseCrucibleResourceCountAndSoftDeleteVolume {
+    conditionally_decrease_references: ConditionallyDecreaseReferences,
+    region_snapshots_to_clean_up_query: RegionSnapshotsToCleanUp,
+    conditionally_update_volume_query: ConditionallyUpdateVolume,
+    volume_id: Uuid,
+}
+
+impl DecreaseCrucibleResourceCountAndSoftDeleteVolume {
+    const UPDATED_REGION_SNAPSHOTS_TABLE: &str = "updated_region_snapshots";
+    const REGION_SNAPSHOTS_TO_CLEAN_UP_TABLE: &str =
+        "region_snapshots_to_clean_up";
+    const UPDATED_VOLUME_TABLE: &str = "updated_volume";
+
+    pub fn new(volume_id: Uuid, snapshot_addrs: Vec<String>) -> Self {
+        Self {
+            conditionally_decrease_references:
+                ConditionallyDecreaseReferences::new(volume_id, snapshot_addrs),
+            region_snapshots_to_clean_up_query: RegionSnapshotsToCleanUp::new(
+                Self::UPDATED_REGION_SNAPSHOTS_TABLE.to_string(),
+            ),
+            conditionally_update_volume_query: ConditionallyUpdateVolume::new(
+                volume_id,
+                Self::REGION_SNAPSHOTS_TO_CLEAN_UP_TABLE.to_string(),
+            ),
+            volume_id,
+        }
+    }
+}
+
+impl QueryId for DecreaseCrucibleResourceCountAndSoftDeleteVolume {
+    type QueryId = ();
+    const HAS_STATIC_QUERY_ID: bool = false;
+}
+
+impl QueryFragment<Pg> for DecreaseCrucibleResourceCountAndSoftDeleteVolume {
+    fn walk_ast<'a>(&'a self, mut out: AstPass<'_, 'a, Pg>) -> QueryResult<()> {
+        use db::schema::volume::dsl;
+
+        out.push_sql("WITH ");
+        out.push_sql(&format!("{} as (", Self::UPDATED_REGION_SNAPSHOTS_TABLE));
+        out.push_sql("UPDATE region_snapshot ");
```

Comment:
Same as I mentioned elsewhere, you should be able to reference the `region_snapshot` table name via:

```rust
out.push_sql("UPDATE ");
db::schema::region_snapshot::table.walk_ast(out.reborrow())?;
```

It's more verbose, but this works even if the table gets renamed

Reply:
done in d242900c197a